README.md
# ascii_chiper
`ascii_chiper` is a Python module for encrypting and decrypting strings, integers, and dictionaries using various encryption techniques. The module provides user-friendly and customizable encryption configurations to suit different use cases and levels of security.

## Features
- Accepts strings, integers, floats, lists and dictionaries as input for encryption
- Supports multiple encryption techniques, including swapping, XOR shifting, interleaving, rotation, XOR base, XOR addition, and interleaving with key
- Offers pre-configured encryption configurations for quick use
- Allows users to create custom encryption configurations

## Installation
To install the ascii_chiper module, use pip: 
```
pip install ascii_chiper
```

## Usage

This module is designed to be simple and easy to use. To get started, import the `Chiper` class and initialize an instance of it. You can then use the `encrypt` and `decrypt` methods to encrypt and decrypt your data.

The `encrypt` method takes the data to be encrypted, a base key, a length, and an encryption configuration as input. You can use one of the predefined encryption configurations or create your own custom configuration. The `decrypt` method requires the encrypted data, base key, length, and the same encryption configuration used during encryption.

### Hello World
```python
from ascii_chiper import Chiper

# Initialize the Chiper
chiper = Chiper(123)

encrypted = chiper.encrypt("Hello World!", 113, 40, Chiper.BASIC_SWAP_INTERLEAVE)
print(f"Encrypted: {encrypted}")
# Output 'xSJJSHNlQWyubF1vDyClV7RvfnKobCZkzyGUIg=='

decrypted = chiper.decrypt(encrypted)
print(f"Decrypted: {decrypted}")
# Output 'Hello World!'
```

### Basic usage
```python
from ascii_chiper import Chiper

# Initialize the Chiper
chiper = Chiper.initialize()

# Encrypt a message using a default encryption configuration
message = "This is a test message."
base, length = 123456789, 5
encrypted_message = chiper.encrypt(message, base, length, encrypt_steps=Chiper.XORBASE_ROTATE)

# Decrypt the encrypted message
decrypted_message = chiper.decrypt(encrypted_message)

print("Original message:", message)
print("Encrypted message:", encrypted_message)
print("Decrypted message:", decrypted_message)
```

### Custom configuration
```python
from ascii_chiper import Chiper, EncryptionModel, DecryptionModel

# Initialize the Chiper
chiper = Chiper.initialize()

# Encrypt a message using a custom encryption configuration
message = "A custom encryption example."
base, length = 987654321, 92
custom_config = [
    {"interleave": {}},
    {"swap": {}},
    {"rotate": {"index": 5}},
    {"xor_base": {"base": 137, "start": 0}},
]

encrypted_message = chiper.encrypt(message, base, length, encrypt_steps=custom_config)

# Decrypt the encrypted message using a DecryptionModel
decryption_model = DecryptionModel.from_encryption_model(EncryptionModel(base, length, custom_config))
decrypted_message = chiper.decrypt(encrypted_message, model=decryption_model)

print("Original message:", message)
print("Encrypted message:", encrypted_message)
print("Decrypted message:", decrypted_message)
```

### Advanced usage
```python
from ascii_chiper import Chiper, KeyGenerator, \
    EncryptionModel, DecryptionModel

# Custom encryption configuration
custom_config = [
    {"interleave": {}},
    {"rotate": {"index": 2}},
    {"xor_add": {"start": 0}},
]

# Create an encryption seed and model
seed = KeyGenerator.generate_seed()
encryption_model = EncryptionModel(123456789, 52, custom_config)

# Encrypt a message using a custom encryption configuration
message = {
    "example": True,
    "message": "This is a test message."
}
encrypted_message = Chiper(seed).encrypt(message, model=encryption_model)

# Decrypt the encrypted message using a DecryptionModel
decryption_model = DecryptionModel.from_encryption_model(encryption_model)
decrypted_message = Chiper(seed).decrypt(encrypted_message, model=decryption_model)

print("Original message:", message)
print("Encrypted message:", encrypted_message)
print("Decrypted message:", decrypted_message)
```

### Custom configuration with lambda functions
```python
from ascii_chiper import Chiper, EncryptionModel, DecryptionModel

# Initialize the Chiper
chiper = Chiper.initialize()

# Encrypt a message using a custom encryption configuration
message = "A custom encryption example."
base, length = 987654321, 92
custom_config = [
    {"swap": {}},
    {"rotate": {
        # We can use a lambda function to define the index
        # the len of the key is passed as the first argument
        "index": lambda key: key % 5,
    }},
    {"xor_base": {
        "base": 137,
        # We can use predefined functions as well
        "start": Chiper.MIDDLE_OF_KEY,
        "end": Chiper.PENULTIMATE_OF_KEY,
    }},
]

encrypted_message = chiper.encrypt(message, base, length, encrypt_steps=custom_config)

# Decrypt the encrypted message using a DecryptionModel
decryption_model = DecryptionModel.from_encryption_model(EncryptionModel(base, length, custom_config))
decrypted_message = chiper.decrypt(encrypted_message, model=decryption_model)

print("Original message:", message)
print("Encrypted message:", encrypted_message)
print("Decrypted message:", decrypted_message)
```

### Decrypt message by knowing key and steps
```python
from ascii_chiper import Chiper, EncryptionModel, DecryptionModel

# Decrypt a message using a custom encryption configuration
encrypted = "2tqBEWERKhGjI3qxKsFLuaMbWhNaybG5s5IqkmrKapJyyypDm6IqklrBmrlrWSq5WyuBkmrKymJqoZoTarErYqOBG6l6yyoiaqLKkqOqOhNaU3paq9ObE5ODKpqTy1pSq4taQ2Oxm1qryyvDe4NrC0q6qwtbWYGLMpIqYnILW9ObyxtiY5FKU3KhaoOrgavTq8taoptZEZJh6hHaEYkRYSMqsaPBerkqG0sTo8lauVqSsZKzyiqSastqQ3KiKpKbwSq5WlmauWsrKpJbmoFTq8F7InqiK1NqksoTMqIqYlLDehOLg0u5e7kqUmKLgYuby1pag5p7g6uTKwtqupsLWqmri1tDOoNCIyrTMsuaU6upG1NyoRqDcstq03qaC2KjuVqSm1lb6ZvqEdphCxGDg0tjCxtLo3N7g3kzI2ERKhGjI3qxKsFLuaMbWhNaybG5s5IqkmrKapJyyypDm6IqklrBmrlrWSq5WyuBkmrKymJqoZoTarErYqOBG6l6yyoiaqLKkqOqOhNaU3paq9ObE5ODKpqTy1pSq4taQ2Oxm1qryyvDe4NrC0q6qwtbWYGLMpIqYnILW9ObyxtiY5FKU3KhaoOrgavTq8taoptZEZJh6hHaK6Ojw4N5MyNhESoRoyN6sSrBS7mjG1oTWsmxubOSKpJqymqScssqQ5uiKpJawZq5a1kquVsrgZKrmntTesErImqiylMykioTUqJ6YovDSxN7gyq5YrmBUpuLWouDy3taq5org2qTmwtauqsLW6k6i0JDKoMyI5rTq8sbU3KpGlNyoWqDessL06OaWmKbuVuSm1kR6erq"
used_encrypt_steps = [
    {"swap": {}},
    {"xor_shift": {}}
]
used_key = [114,114]
excepted_output = [["0","Edt6O8E7ictbK9K76RvREYMRMyNhETsRE8K7S+m7EekR0YMLY4MbS6MLe0t5cyODETMRYUtbGjKKOzubsprSEyrJKquhK6lKsyuxepoamWIau+kq0RFLEaNzkysLc2ljI4NpM0uzuyuTK+sR"],["1","Edt6O8E7ictbK9K76RvREYMRMyNhETsRE8K7S+m7EekR0Sujo8ODeTMjYRFbETJLOxqbipo7E7LJ0qsqKypKoSuperMasWKau5kqGhHpEdFzSyujc5NjC4NpMyOzaStLK7sRk+s="],["application/pdf","Edt6O8E7ictbK9K76RvREYMRMyNhETsRE8K7S+m7EekR0YMLY4MbS6MLe0t5cyODETMRYUtbGjKKOzubsprSEyrJKquhK6lKsyuxepoamWIau+kq0RFLEaNzkysLc2ljI4NpM0uzuyuTK+sR"],["text/pdf","Edt6O8E7ictbK9K76RvREYMRMyNhETsRE8K7S+m7EekR0Sujo8ODeTMjYRFbETJLOxqbipo7E7LJ0qsqKypKoSuperMasWKau5kqGhHpEdFzSyujc5NjC4NpMyOzaStLK7sRk+s="]]

# Create the EncryptionModel
enc_model = EncryptionModel(0, 0, encrypt_steps=used_encrypt_steps)

# Initialize the Chiper
chiper = Chiper.initialize()

# Decrypt the encrypted message
decrypted_message = chiper.decrypt(encrypted, key=used_key, decrypt_steps=DecryptionModel.from_encryption_model(enc_model).decrypt_steps)
print("Decrypted message:", decrypted_message)
print(f"Decrypted message is equal to excepted output: {decrypted_message == excepted_output}")
# Have fun :P those looks like some mime_types encrypted
```
For other examples usages, please refer to the Examples folder.

## Encryption Methods

`ascii_chiper` offers various encryption techniques that can be combined in different configurations to achieve the desired level of security:

| Name            | Description                                          | Optional Parameters                 |
|-----------------|------------------------------------------------------|-------------------------------------|
| reverse         | Reverse the input data.                              |                                     |
| swap            | Swaps pairs of characters in the input data.         |                                     |
| rotate          | Rotates the input data by a specified index from the key. | Index from key to use           |
| interleave      | Interleaves the input data, effectively rearranging the characters. | key start/end                 |
| interleave_key  | Interleaves the input data based on the key start and end positions. | key start/end                 |
| xor_shift       | Applies an XOR shift operation using a specified index from the key. | index of key to use           |
| xor_base        | Performs XOR operation on the input data using the base and the key start and end positions. | base to use, start/end key to use |
| xor_add         | Adds the key start and end values to the input data using XOR addition. | start/end key to use         |

You can combine these encryption methods in a custom configuration to suit your specific requirements.

## Encryption Configurations
In addition to creating your own custom encryption configurations, `ascii_chiper` provides several pre-configured encryption configurations that cater to different use cases and security levels:

- **BASIC_SWAP_INTERLEAVE**: A simple configuration that combines Swap and Interleave encryption methods for basic security.
- **ROTATE_XORSHIFT**: Combines the Rotate and XorShift methods for an intermediate level of security.
- **XORBASE_ROTATE**: Combines the XorBase and Rotate methods for enhanced security.
- **XORADD_INTERLEAVE**: Combines the XorAdd and Interleave methods for a higher level of security.
- **FULL_ENCRYPTION**: Uses all available encryption methods for maximum security.

Remember to use the same encryption configuration for both encryption and decryption to ensure proper functionality.

## Personal Thoughts

I hope you find this module useful. This module is the exact Python implementation of the encryption method used by a well-known web anti-bot system. If you know what I'm talking about, you'll definitely find it useful. 
Please feel free to contact me for any help or suggestions via [Email](mailto:glizzykingdreko@protonmail.com) or [Twitter](https://mobile.twitter.com/glizzykingdreko). I appreciate your feedback and contributions to the project.

## License
This project is licensed under the MIT [License](LICENSE). See the LICENSE file for more details.

## My links
- [Project repository](https://github.com/glizzykingdreko/ascii_chiper)
- [GitHub](https://github.com/glizzykingdreko)
- [Twitter](https://mobile.twitter.com/glizzykingdreko)
- [Medium](https://medium.com/@glizzykingdreko)
- [Email](mailto:glizzykingdreko@protonmail.com)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.1.4] - 2023-06-11

### Added
- Added `GeneratorHelper` to handle support for both MacOS and Raspberry Pi.

## [0.1.3] - 2023-03-25

### Fixed

- Fixed a minor issue to accept both floats and integers as input.

## [0.1.2] - 2023-03-25

### Added
- Added optional start and end key arguments for the Interleave encryption method.
- Support for lambda functions in the encryption/decryption steps options (index/start/end). The key length will be passed as an argument.
- Added premade lambda functions: `Chiper.PENULTIMATE_OF_KEY` and `Chiper.MIDDLE_OF_KEY`.
- Implemented type checking for input parameters.

### Changed
- Modified the encrypt/decrypt steps from `Dict[str, dict]` to `List[Dict[str, dict]]` to support using the same encryption method multiple times in the sequence.

### Fixed

### Removed

## [0.1.1] - 2023-03-23
### Added
- Added the possibility to encrypt/decrypt by passing an encryption key directly.
- Added "reverse" as an encryption and decryption method.
- Improved documentation.

### Changed

### Fixed
- Fixed swap reverse issue.

### Removed

## [0.1.0] - 2023-03-22
### Added
- Initial implementation of `ascii_chiper`.
- Support for encrypting and decrypting strings, integers, and dictionaries.
- Multiple encryption techniques including swapping, XOR shifting, interleaving, rotation, XOR base, XOR addition, and interleaving with key.
- Pre-configured encryption configurations for quick use.
- Custom encryption configuration support.
- Examples demonstrating various use cases of the module.
- Detailed README with usage instructions, encryption methods, and configurations.
- MIT License and Personal Thoughts section in the README.

### Changed

### Fixed

### Removed


ascii_chiper/chiper.py
from typing import Dict, Union, Any, List, Tuple

from .key_generator import KeyGenerator
from .exceptions import InvalidModeException, \
    InvalidStartIndexException, InvalidEndIndexException, InvalidKeyInputException, \
        EncryptionException, DecryptionException
from .models import DecryptionModel, EncryptionModel
from .utils import swap_back, swap, xor_unshift, xor_shift, \
    deinterleave, interleave, unrotate, rotate, unxor_base, xor_base, xor_unadd, xor_add, \
        deinterleave_key, interleave_key, string_to_ascii, ascii_to_string, ascii_to_base64, \
            base64_to_ascii, clean_input, revert_clean_input, reverse

class Chiper:
    """Encrypts and decrypts messages using a key"""

    BASIC_SWAP_INTERLEAVE = [
        {"interleave": {}},
        {"swap": {}},
    ]

    ROTATE_XORSHIFT = [
        {"rotate": {"index": 5}},
        {"xor_shift": {"index": 7}},
    ]

    XORBASE_ROTATE = [
        {"xor_base": {"base": 113, "start": 0}},
        {"rotate": {"index": 3}},
    ]

    XORADD_INTERLEAVE = [
        {"xor_add": {"start": 0}},
        {"interleave": {}},
    ]

    FULL_ENCRYPTION = [
        {"interleave": {}},
        {"swap": {}},
        {"xor_shift": {"index": 0}},
        {"rotate": {"index": 0}},
        {"xor_base": {"base": 113, "start": 0}},
        {"interleave_key": {"start": 0}},
        {"reverse": {}},
    ]
    PENULTIMATE_OF_KEY = lambda k: k-1
    MIDDLE_OF_KEY = lambda k: int((k-1)/2)


    @staticmethod
    def initialize() -> "Chiper":
        """Initializes a Chiper object with a random key.

        Returns:
            Chiper: A Chiper object with a random key.
        """
        return Chiper(KeyGenerator.generate_seed())
    
    @staticmethod
    def encrypt_from_model(message: str, seed: int, model: Union[EncryptionModel, DecryptionModel]) -> str:
        """Encrypts a message using a model.

        Args:
            message: The message to encrypt.
            seed: The seed to use for key generation.
            model: The model to use for encryption.

        Returns:
            str: The encrypted message.
        """
        return Chiper(seed).encrypt(message, *model())
    
    @staticmethod
    def format_step_params(step_params: Dict[str, Any], key: List[int]) -> Tuple:
        """Formats the parameters for a step.

        Args:
            step_params: The parameters to format.
            key: The key to use for formatting.

        Returns:
            Dict[str, Any]: The formatted parameters.
        """
        index, start, end, base = step_params.get('index', 0), step_params.get('start', 0), \
            step_params.get('end', len(key)), step_params.get('base', 113)
        if callable(index): index = index(len(key))
        if callable(start): start = start(len(key))
        if callable(end): end = end(len(key))
        if callable(base): base = base(len(key)) 
        if start == "PENULTIMATE_OF_KEY": start = Chiper.PENULTIMATE_OF_KEY(len(key))
        if start == "MIDDLE_OF_KEY": start = Chiper.MIDDLE_OF_KEY(len(key))
        if end == "PENULTIMATE_OF_KEY": end = Chiper.PENULTIMATE_OF_KEY(len(key))
        if end == "MIDDLE_OF_KEY": end = Chiper.MIDDLE_OF_KEY(len(key))
        if base == "PENULTIMATE_OF_KEY": base = Chiper.PENULTIMATE_OF_KEY(len(key))
        if base == "MIDDLE_OF_KEY": base = Chiper.MIDDLE_OF_KEY(len(key))
        if start < 0 or start >= len(key):
            raise InvalidStartIndexException(f"Invalid start index: {start}")
        if end < 0 or end > len(key):
            raise InvalidEndIndexException(f"Invalid end index: {end}")
        return index, start, end, base
    
    @staticmethod
    def check_inputs_types(key: List[int], base: int, len: int, steps: List[Dict[str, Any]], message: Union[str, int, Dict]) -> None:
        """Checks the types of the inputs.

        Args:
            key: The key to check.
            base: The base to check.
            len: The length of the key to check.
            steps: The steps to check.
            message: The message to check.
        """
        if base is not False and not isinstance(base, int):
            raise InvalidKeyInputException("Invalid base input: base must be an integer.")
        
        if len is not False and not isinstance(len, int):
            raise InvalidKeyInputException("Invalid len input: len must be an integer.")
        
        if not isinstance(steps, list) or not all(isinstance(step, dict) and isinstance(next(iter(step.keys())), str) and isinstance(next(iter(step.values())), dict) for step in steps):
            raise InvalidKeyInputException("Invalid steps input: steps must be a list of dictionaries in the format {str, dict}.")
        
        if not isinstance(message, (str, int, dict, list, float)):
            raise InvalidKeyInputException("Invalid message input: message must be a string, integer, float, list or dictionary.")

    def __init__(self, seed: int):
        self.seed, self.encryption_model, self.plain_text, self.base, \
            self.lenght, self.used_key, self.decrypt_model= \
                seed, False, "", 0, 0, [], False

    def encrypt(
        self, 
        message: Union[str, dict, int]=False, 
        base: int=False, 
        lenght: int=0, 
        encrypt_steps: List[Dict[str, Dict]]=False, 
        model: EncryptionModel=False,
        key: List[int]=False,
    ) -> str:
        """
        ### Encrypts a message using a key and a set of steps or a model.

        Args:
            `message` (Union[str, dict, int]): The message to encrypt. Default is False.
            `base` (int): The base for key generation. Default is False.
            `lenght` (int): The length of the key. Default is 0.
            `encrypt_steps` (List[Dict[str, Dict]]): The steps to use for encryption. Default is False.
            `model` (EncryptionModel): The model to use for encryption. Default is False.
            `key` (List[int]): The key to use for encryption. Default is False. If False, a key will be generated.
        
        Returns:
            str: The encrypted message.

        Raises:
            InvalidModeException: If the mode is invalid or missing/invalid arguments.
            InvalidKeyInputException: If one of the input keys is invalid.
            ValueError: If the message is missing.
            InvalidModeException: If the mode is invalid.
            EncryptionException: If the encryption fails.
        
        Examples:
            >>> from ascii_chiper import Chiper
            >>> chiper = Chiper(123)
            >>> chiper.encrypt("Hello World!", 113, 40, Chiper.BASIC_SWAP_INTERLEAVE)
            'xSJJSHNlQWyubF1vDyClV7RvfnKobCZkzyGUIg=='
        """
        try:
            if isinstance(message, bool):
                if not self.plain_text:
                    raise ValueError("Missing message")
                message = self.plain_text
            if model:
                base, lenght, encrypt_steps = model()
            elif not encrypt_steps or not base or not lenght:
                if self.decrypt_model:
                    base, lenght, encrypt_steps = EncryptionModel.from_decryption_model(self.decrypt_model)()
                elif not key:
                    raise ValueError("Missing arguments")
        except: raise InvalidModeException("Invalid mode or missing/invalid arguments")
        if not key:
            try: key = KeyGenerator(self.seed).create_key(base, lenght)
            except: raise InvalidKeyInputException("Invalid key input")
        elif (not isinstance(key, list) or not all(isinstance(k, int) for k in key)):
            raise InvalidKeyInputException("Invalid key input: key must be a list of integers.")
        
        Chiper.check_inputs_types(key, base, lenght, encrypt_steps, message)
        try:
            self.encryption_model = EncryptionModel(base, lenght, encrypt_steps)
            ascii_list = string_to_ascii(clean_input(message))
            for item in encrypt_steps:
                step_name, step_params = next(iter(item.items()))

                # Initialize/parse/execute the options
                index, start, end, base = Chiper.format_step_params(step_params, key)

                # Execute the step
                if step_name == 'reverse':
                    ascii_list = reverse(ascii_list)
                elif step_name == 'swap':
                    ascii_list = swap(ascii_list)
                elif step_name == 'xor_shift':
                    ascii_list = xor_shift(ascii_list, key, index)
                elif step_name == 'rotate':
                    ascii_list = rotate(ascii_list, key, index)
                elif step_name == 'xor_base':
                    ascii_list = xor_base(ascii_list, key, base, start, end)
                elif step_name == 'xor_add':
                    ascii_list = xor_add(ascii_list, key, start, end)
                elif step_name == 'interleave':
                    ascii_list = interleave(ascii_list, key, start, end)
                elif step_name == 'interleave_key':
                    ascii_list = interleave_key(ascii_list, key, start, end)
                else: raise InvalidModeException(f"Invalid mode: {step_name}")
            
            # Save the encryption data
            self.plain_text, self.base, self.lenght, self.used_key, self.encrypt_steps = \
                message, base, lenght, key, encrypt_steps
            
            # Return the encrypted message
            return ascii_to_base64(ascii_list)
        except:
            raise EncryptionException("Encryption failed")
        

    def decrypt(
        self, 
        message: Union[str, dict, int]=False, 
        base: int=False, 
        lenght: int=0,
        decrypt_steps: Dict[str, Any]=False, 
        model: DecryptionModel=False,
        key: List[int]=False
    ) -> Union[str, dict, int]:
        """
        ### Decrypts a message using a key and a set of steps or a model.

        Args:
            `message` (Union[str, dict, int]): The message to decrypt. Default is False.
            `base` (int): The base for key generation. Default is False.
            `lenght` (int): The length of the key. Default is 0.
            `decrypt_steps` (Dict[str, Any]): The steps to use for decryption. Default is False.
            `model` (DecryptionModel): The model to use for decryption. Default is False.
            `key` (List[int]): The key to use for decryption. Default is False. If False, a key will be generated.
        
        Returns:
            The decrypted message.
        
        Raises:
            InvalidModeException: If the mode is invalid or missing/invalid arguments.
            InvalidKeyInputException: If one of the input keys is invalid.
            ValueError: If the message is missing.
            InvalidModeException: If the mode is invalid.
            DecryptionException: If the decryption failed.
        
        Examples:
            >>> from ascii_chiper import Chiper
            >>> chiper = Chiper(123)
            >>> encrypted = chiper.encrypt("Hello World!", 113, 40, Chiper.BASIC_SWAP_INTERLEAVE)
            >>> chiper.decrypt(encrypted)
            'Hello World!'
        """
        try:
            if not message:
                if not self.plain_text:
                    raise ValueError("Missing message")
                message = self.plain_text
            if model:
                base, lenght, decrypt_steps = model()
            elif not decrypt_steps or not base or not lenght:
                if self.encryption_model:
                    base, lenght, decrypt_steps = DecryptionModel.from_encryption_model(self.encryption_model)()
                elif not key:
                    raise ValueError("Missing arguments")
        except: raise InvalidModeException("Invalid mode or missing/invalid arguments")
        if not key:
            try: key = KeyGenerator(self.seed).create_key(base, lenght)
            except: raise InvalidKeyInputException("Invalid key input")
        elif (not isinstance(key, list) or not all(isinstance(k, int) for k in key)):
            raise InvalidKeyInputException("Invalid key input: key must be a list of integers.")
        
        Chiper.check_inputs_types(key, base, lenght, decrypt_steps, message)
        decrypt_steps = list(reversed(decrypt_steps))
        try:
            self.decrypt_model = DecryptionModel(base, lenght, decrypt_steps)
            ascii_list = base64_to_ascii(message)
            for item in decrypt_steps:
                step_name, step_params = next(iter(item.items()))

                # Initialize/parse/execute the options
                index, start, end, base = Chiper.format_step_params(step_params, key)
                
                # Execute the step
                if step_name == 'reverse':
                    ascii_list = reverse(ascii_list)
                elif step_name == 'swap_back':
                    ascii_list = swap_back(ascii_list)
                elif step_name == 'xor_unshift':
                    ascii_list = xor_unshift(ascii_list, key, index)
                elif step_name == 'unrotate':
                    ascii_list = unrotate(ascii_list, key, index)
                elif step_name == 'unxor_base':
                    ascii_list = unxor_base(ascii_list, key, base, start, end)
                elif step_name == 'xor_unadd':
                    ascii_list = xor_unadd(ascii_list, key, start, end)
                elif step_name == 'deinterleave':
                    ascii_list = deinterleave(ascii_list, key)
                elif step_name == 'deinterleave_key':
                    ascii_list = deinterleave_key(ascii_list, key, start, end)
                else: raise InvalidModeException(f"Invalid mode: {step_name}")
            
            # Save the decryption data
            self.plain_text, self.base, self.lenght, self.used_key, self.decrypt_steps = \
                message, base, lenght, key, decrypt_steps
            
            # Return the decrypted message
            return revert_clean_input(ascii_to_string(ascii_list))
        except:
            raise DecryptionException("Decryption failed")